sorrydb/agents/json_agent.py

```
@@ -0,0 +1,137 @@
+        json.JSONDecodeError: If the JSON file is invalid
+    """
+    logger.info(f"Loading sorry JSON from {json_path}")
+    try:
+        with open(json_path, "r") as f:
+            sorry_data = json.load(f, object_hook=sorry_object_hook)
+
+        return sorry_data
+    except FileNotFoundError:
+        logger.error(f"Sorry JSON file not found: {json_path}")
+        raise
+    except json.JSONDecodeError:
+        logger.error(f"Invalid JSON in sorry file: {json_path}")
+        raise
+
+
+def save_proofs_json(output_path: Path, output: List[Dict]):
+    """Save the proofs to a JSON file.
+
+    Args:
+        output_path: Path to output JSON file
+        output: list of dicts with sorries and proofs
+    """
+    try:
+        with open(output_path, "w") as f:
+            json.dump(output, f, indent=4, cls=SorryJSONEncoder)
+    except Exception as e:
+        logger.error(f"Error saving proofs to {output_path}: {e}")
+        raise
+
+
+class SorryStrategy(Protocol):
+    def prove_sorry(self, repo_path: Path, sorry: Sorry) -> str | None:
+        """ To be implemented by the agent
+        Args:
+            repo_path: Path to the repository
+            sorry: sorry to prove
+
+        Returns:
+            Proof string to replace "sorry" or None if no proof was found
+        """
+        pass
+
+class JsonAgent:
+    def __init__(self, strategy: SorryStrategy, lean_data_path: Path | None = None):
+        self.strategy = strategy
+        self.lean_data_path = lean_data_path
+
+    def _process_sorries(self, local_sorries: list[Sorry], lean_data_dir: Path) -> list[dict]:
+        proofs = []
+        for sorry in local_sorries:
+            # Prepare the repository (clone and checkout)
+            checkout_path = prepare_repository(
+                sorry.repo.remote,
+                sorry.repo.branch,
+                sorry.repo.commit,
+                lean_data_dir,
+            )
+            if not checkout_path:
```

This shouldn't be fixed in this PR. However, I have created an #129 to raise expectations instead of returning None in `prepare_repository`.

---

Agreed! This seems to be a duplicate of #122